File: celest/src/main/java/be/angelcorp/libs/celest/frames/ReferenceFrameGraph.java
/**
 * Copyright (C) 2009-2012 simon <simon@angelcorp.be>
 *
 * Licensed under the Non-Profit Open Software License version 3.0
 * (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 *        http://www.opensource.org/licenses/NOSL3.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package be.angelcorp.libs.celest.frames;

import java.util.Iterator;
import java.util.List;
import java.util.Set;

import org.jgrapht.WeightedGraph;
import org.jgrapht.alg.DijkstraShortestPath;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import be.angelcorp.libs.celest.time.Epoch;

import com.google.common.base.Predicate;
import com.google.common.collect.Iterators;

/**
 * Implementation of the {@link IReferenceFrameGraph}. This is the graph that contains all the reference
 * frames, and the possible transforms between them.
 * 
 * @author Simon Billemont
 */
public class ReferenceFrameGraph implements IReferenceFrameGraph {

	/**
	 * Predicate to locate the exact instance of a {@link IReferenceFrame} in the
	 * {@link IReferenceFrameGraph}.
	 * 
	 * @param frame
	 *            Instance of an {@link IReferenceFrame} to locate.
	 * @return Predicate to search for the instance.
	 */
	public static Predicate<IReferenceFrame> exactFrame(final IReferenceFrame frame) {
		return new Predicate<IReferenceFrame>() {
			@Override
			public boolean apply(IReferenceFrame input) {
				return input == frame;
			}
		};
	}

	/** This is the JGraphT that actually describes the {@link IReferenceFrame}'s and there connections */
	private final WeightedGraph<IReferenceFrame, IReferenceFrameTransformFactory<?, ?>>	graph;

	private static final Logger															logger	= LoggerFactory
																										.getLogger(ReferenceFrameGraph.class);

	/**
	 * Create a new empty {@link IReferenceFrameGraph} using a specified JGraphT instance as graph
	 * back-end.
	 * 
	 * @param graph
	 *            Graph to use as back-end.
	 */
	public ReferenceFrameGraph(WeightedGraph<IReferenceFrame, IReferenceFrameTransformFactory<?, ?>> graph) {
		this.graph = graph;
	}

	/***
	 * {@inheritDoc}
	 */
	@Override
	public void attachFrame(IReferenceFrame frame) {
		graph.addVertex(frame);
	}

	/***
	 * {@inheritDoc}
	 */
	@Override
	public <F1 extends IReferenceFrame, F2 extends IReferenceFrame> void attachTransform(F1 frame1, F2 frame2,
			IReferenceFrameTransformFactory<F1, F2> transform) {
		if (!graph.containsVertex(frame1)) {
			logger.debug("Tried to add transform between frame {} and {}, but frame {} does not exist in the graph",
					new Object[] { frame1, frame2, frame1 });
			return;
		}
		if (!graph.containsVertex(frame2)) {
			logger.debug("Tried to add transform between frame {} and {}, but frame {} does not exist in the graph",
					new Object[] { frame1, frame2, frame2 });
			return;
		}
		graph.addEdge(frame1, frame2, transform);
	}

	/**
	 * Find the path between two {@link IReferenceFrame} instance that should be in the graph.
	 * 
	 * @param from
	 *            Origin of path.
	 * @param to
	 *            Destination of the path.
	 * @return A path describing all the nodes to visit (in sequence) that lead from the origin to the
	 *         destination.
	 */
	protected List<IReferenceFrameTransformFactory<?, ?>> findPath(IReferenceFrame from, IReferenceFrame to) {
		DijkstraShortestPath<IReferenceFrame, IReferenceFrameTransformFactory<?, ?>> alg =
				new DijkstraShortestPath<>(graph, from, to);
		return alg.getPathEdgeList();
	}

	/**
	 * Find the path between two {@link IReferenceFrame} instance that should be in the graph.
	 * <p>
	 * Note: the first
	 * </p>
	 * 
	 * @param from
	 *            Predicate to find the origin of path.
	 * @param to
	 *            Predicate to find the destination of the path.
	 * @return A path describing all the nodes to visit (in sequence) that lead from the origin to the
	 *         destination.
	 * @throws ReferenceFrameTransformationException
	 *             When no match was found for either the origin or destination frame.
	 */
	protected List<IReferenceFrameTransformFactory<?, ?>> findPath(Predicate<IReferenceFrame> from,
			Predicate<IReferenceFrame> to) throws ReferenceFrameTransformationException {
		Set<IReferenceFrame> all_frames = graph.vertexSet();

		IReferenceFrame to_instance = null, from_instance = null;
		boolean from_found = false, to_found = false;
		for (IReferenceFrame iReferenceFrame : all_frames) {
			if (!from_found && from.apply(iReferenceFrame)) {
				from_instance = iReferenceFrame;
				from_found = true;
			}
			if (!to_found && to.apply(iReferenceFrame)) {
				to_instance = iReferenceFrame;
				to_found = true;
			}
			if (from_found && to_found)
				break;
		}

		if (from == null || to == null)
			throw new ReferenceFrameTransformationException(
					"Could not find any frame matching the from or to frame predicate. " +
							"Did you forget to attach the frame to the framegraph?");

		return findPath(from_instance, to_instance);
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public IReferenceFrame findReferenceFrame(Predicate<IReferenceFrame> frame_predicate) {
		IReferenceFrame frame = null;
		for (IReferenceFrame iReferenceFrame : graph.vertexSet()) {
			if (frame_predicate.apply(iReferenceFrame)) {
				frame = iReferenceFrame;
				break;
			}
		}
		return frame;
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public Iterator<IReferenceFrameTransformFactory<?, ?>> findReferenceFrameTransforms(final IReferenceFrame frame) {
		return new Iterator<IReferenceFrameTransformFactory<?, ?>>() {

			/** Flag indicating that no more elements can be retrieved */
			private boolean											hasNext;
			/** The next element returned by {@link Iterator#next()} */
			private IReferenceFrameTransformFactory<?, ?>			next;
			/** Iterator to iterator over the remaining transforms possibly connected to the given frame */
			final Iterator<IReferenceFrameTransformFactory<?, ?>>	transforms	= graph.edgeSet().iterator();

			{
				// Upon construction, retrieve the next element
				retrieveNext();
			}

			@Override
			public boolean hasNext() {
				return hasNext;
			}

			@Override
			public IReferenceFrameTransformFactory<?, ?> next() {
				IReferenceFrameTransformFactory<?, ?> tmp = next;
				retrieveNext();
				return tmp;
			}

			@Override
			public void remove() {
				throw new UnsupportedOperationException();
			}

			private void retrieveNext() {
				do {
					if (transforms.hasNext()) {
						IReferenceFrameTransformFactory<?, ?> transform = transforms.next();
						if (graph.getEdgeSource(transform) == frame || graph.getEdgeTarget(transform) == frame) {
							next = transform;
							break;
						}
					} else {
						hasNext = false;
						break;
					}
				} while (true);
			}
		};
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public Iterator<IReferenceFrameTransformFactory<?, ?>> findReferenceFrameTransforms(
			Predicate<IReferenceFrame> frame_predicate) {
		// Locate the frame matching the predicate
		IReferenceFrame frame = findReferenceFrame(frame_predicate);

		// If no frame was found, return an empty iterator
		if (frame == null)
			return Iterators.emptyIterator();

		// Else return all the connected IReferenceFrameTransformFactories
		return findReferenceFrameTransforms(frame);
	}

	/**
	 * Get the reference to the JGraphT back-end of this {@link IReferenceFrameGraph}.
	 * 
	 * @return JGraphT instance of this reference frame.
	 */
	public WeightedGraph<IReferenceFrame, IReferenceFrameTransformFactory<?, ?>> getJGraphT() {
		return graph;
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public Iterator<IReferenceFrame> getReferenceFrames() {
		return graph.vertexSet().iterator();
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public IReferenceFrameTransform<?, ?> getTransform(Predicate<IReferenceFrame> from, Predicate<IReferenceFrame> to,
			Epoch epoch) throws ReferenceFrameTransformationException {
		// Get the respective factory for the given input
		IReferenceFrameTransformFactory<?, ?> factory = getTransformFactory(from, to);
		// Create a new transform for the specific epoch
		IReferenceFrameTransform<?, ?> transform = factory.getTransform(epoch);

		return transform;
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public <T extends IReferenceFrame, V extends IReferenceFrame> IReferenceFrameTransform<T, V>
			getTransform(T from, V to, Epoch epoch) throws ReferenceFrameTransformationException {
		// Get the respective factory for the given input
		IReferenceFrameTransformFactory<T, V> factory = getTransformFactory(from, to);
		// Create a new transform for the specific epoch
		IReferenceFrameTransform<T, V> transform = factory.getTransform(epoch);

		return transform;
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public IReferenceFrameTransformFactory<?, ?> getTransformFactory(Predicate<IReferenceFrame> from,
			Predicate<IReferenceFrame> to) throws ReferenceFrameTransformationException {
		// Find the path between the two first matching frames
		List<IReferenceFrameTransformFactory<?, ?>> path = findPath(from, to);

		// Make sure a path exits
		if (path == null)
			throw new ReferenceFrameTransformationException(
					"No ReferenceFrame transformation path exists between frame '%s' and '%s'", from, to);

		// Create the respective factory from the path
		IReferenceFrameTransformFactory<?, ?> factory = pathToTransformFactory(path);

		return factory;
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public <T extends IReferenceFrame, V extends IReferenceFrame> IReferenceFrameTransformFactory<T, V> getTransformFactory(
			T from, V to) throws ReferenceFrameTransformationException {
		// Find the path between the two first matching frames
		List<IReferenceFrameTransformFactory<?, ?>> path = findPath(from, to);

		// Make sure a path exits
		if (path == null)
			throw new ReferenceFrameTransformationException(
					"No ReferenceFrame transformation path exists between frame '%s' and '%s'", from, to);

		// Create the respective factory from the path
		IReferenceFrameTransformFactory<?, ?> factory = pathToTransformFactory(path);

		return (IReferenceFrameTransformFactory<T, V>) factory;
	}

	/**
	 * Create an {@link IReferenceFrameTransformFactory} from a known path over various compatible
	 * {@link IReferenceFrame} nodes in the frame graph.
	 * 
	 * @param path
	 *            The path leading from one specific frame to another frame.
	 * @return An {@link IReferenceFrameTransformFactory} to convert a state in the origin frame to the
	 *         destination frame.
	 */
	protected IReferenceFrameTransformFactory<?, ?> pathToTransformFactory(
			List<IReferenceFrameTransformFactory<?, ?>> path) {

		Iterator<IReferenceFrameTransformFactory<?, ?>> it = path.iterator();
		IReferenceFrameTransformFactory<?, ?> currentFactory = it.next();

		while (it.hasNext()) {
			IReferenceFrameTransformFactory newFactory = it.next();
			currentFactory = currentFactory.add(newFactory);
		}

		return currentFactory;
	}

}

File: celest/src/main/java/be/angelcorp/celest/frameGraph/ReferenceFrameGraph.scala
/**
 * Copyright (C) 2009-2012 simon <simon@angelcorp.be>
 *
 * Licensed under the Non-Profit Open Software License version 3.0
 * (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 * http://www.opensource.org/licenses/NOSL3.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package be.angelcorp.celest.frameGraph

import be.angelcorp.celest.time.Epoch

/**
 * Abstract description a collection of interlinked frameGraph.
 * <p>
 * The ReferenceFrameGraph is used to find transforms between any reference frame that was
 * registered to it. This way, uses do not need to know all the individual transforms required to
 * transform between any of the frameGraph, even if they require multiple complex transformations.
 * </p>
 * <p>
 * There are two flavors to get a ReferenceFrameTransformFactory or ReferenceFrameTransform. The first
 * is when the actual source/destination frame instance is known, the other is using a predicate to
 * identify the specific frameGraph.
 * </p>
 *
 * <p>
 * Note: The first frame to match either the origin/destination predicate is used.
 * </p>
 *
 * @author Simon Billemont
 */
trait ReferenceFrameGraph {

  /**
   * Attach a new frame instance to this frame graph. If the frame already exists, no duplicate of the
   * frame is attached (the same instance can only be attached once).
   *
   * @param frame Frame to attach to this graph.
   */
  def attachFrame(frame: ReferenceSystem)

  /**
   * Attach a transform between two existing ReferenceFrames in this graph.
   * <p>
   * If either frame does not exist in the frame graph, no action is performed.<br/>
   * If the same transform instance was already connected to the graph, its origin/destination frame
   * will be replaced by the passed frameGraph.
   * </p>
   *
   * @param frame1    Origin reference frame of the transformation.
   * @param frame2    Destination reference frame of the transformation.
   * @param transform Transformation factory to to produce transforms between the origin and destination frame.
   */
  def attachTransform[F1 <: ReferenceSystem, F2 <: ReferenceSystem](frame1: F1, frame2: F2, transform: ReferenceFrameTransformFactory[F1, F2])

  /**
   * Find the first frame from all the connected ReferenceFrames that match a specified
   * predicate. If none of the frameGraph match, null is returned.
   *
   * @param frame_predicate Predicate to identify a specific frame.
   * @return The first frame to match the given predicate or else null.
   */
  def findReferenceFrame(frame_predicate: ReferenceSystem => Boolean): Option[ReferenceSystem]

  /**
   * Search for all the ReferenceFrameTransformFactories connected to the provided frame
   * instance. If the frame was not attached to this graph, then an empty iterator is returned.
   *
   * @param frame Find all factories to connected to this frame instance.
   * @return Iterator over all connected factories.
   */
  def findReferenceFrameTransforms(frame: ReferenceSystem): Iterable[ReferenceFrameTransformFactory[_, _]]

  /**
   * Searches for the first frame to match the given predicate. After this, an iterator over all the
   * IReferenceFrameTransformFactories connected to the found frame are returned. If the frame
   * was not found in this graph, then an empty iterator is returned.
   *
   * @param frame_predicate Find all factories to connected to the first matching frame with this predicate.
   * @return Iterator over all connected factories.
   */
  def findReferenceFrameTransforms(frame_predicate: ReferenceSystem => Boolean): Iterable[ReferenceFrameTransformFactory[_, _]]

  /**
   * Returns an iterator that iterates over all the registered ReferenceFrames.
   *
   * @return Iterator for all the frameGraph in this graph.
   */
  def getReferenceFrames: Iterable[ReferenceSystem]

  /**
   * Find a transformation between two ReferenceFrames. The source and destination ReferenceFrame are identified
   * using a predicate, where the the first frame to match the predicate is used as either source or destination
   * frame respectively. The generated transformation is valid for the specified epoch.
   *
   * @param from  Predicate to find the origin reference frame.
   * @param to    Predicate to find the destination reference frame.
   * @param epoch The epoch at which the ReferenceFrameTransform must be valid.
   * @return An ReferenceFrameTransform describing the transformation between the first matched frame of the "from" and "to" predicates.
   */
  def getTransform(from: ReferenceSystem => Boolean, to: ReferenceSystem => Boolean, epoch: Epoch): Option[ReferenceFrameTransform[_, _]]

  /**
   * Find a transformation between two known ReferenceFrames.
   *
   * <p>
   * Note: Both frameGraph need to be attached to this reference frame graph, and a transformation chain
   * must exist between the two frameGraph in order to yield a valid resulting transformation.
   * </p>
   *
   * @param from  Origin reference frame.
   * @param to    Destination reference frame.
   * @param epoch The epoch at which the ReferenceFrameTransform must be valid.
   * @return An ReferenceFrameTransform describing the transformation between the two provided reference frameGraph.
   */
  def getTransform[F <: ReferenceSystem, T <: ReferenceSystem](from: F, to: T, epoch: Epoch): Option[ReferenceFrameTransform[F, T]]

  /**
   * Find a factory capable of generating transformations between two ReferenceFrames. The source and destination
   * ReferenceFrame are identified using a predicate, where the the first frame to match the predicate is used as
   * either source or destination frame respectively. The generated transformation is valid for the specified epoch.
   *
   * @param from  Predicate to find the origin reference frame.
   * @param to    Predicate to find the destination reference frame.
   * @return An ReferenceFrameTransformFactory capable of producing transformations between the first matched frame of
   *         the "from" and "to" predicates.
   */
  def getTransformFactory(from: ReferenceSystem => Boolean, to: ReferenceSystem => Boolean): Option[ReferenceFrameTransformFactory[_, _]]

  /**
   * Find a factory capable of producing transformations between two known ReferenceFrames.
   *
   * <p>
   * Note: Both frameGraph need to be attached to this reference frame graph, and a transformation chain
   * must exist between the two frameGraph in order to yield a valid transformation factory.
   * </p>
   *
   * @param from  Origin reference frame.
   * @param to    Destination reference frame.
   * @return A ReferenceFrameTransformFactory capable of producing transformations between the two provided reference frameGraph.
   */
  def getTransformFactory[F <: ReferenceSystem, T <: ReferenceSystem](from: F, to: T): Option[ReferenceFrameTransformFactory[F, T]]

}
